### DISCUSSION TITLE
perPage limit above 500

### DISCUSSION THREAD

**Discussion by dindinet** [pocketbase/pocketbase#2026] [Q&A]
Is it possible to set the perPage higher than the current max of 500

api/collections/posts/records?perPage=1000 will only allow max of 500

I want to use the api call to fetch content for hugo build and paging would be a pain.

Thanks in advance

**Comment by ganigeorgiev** [comment]
At the moment the `MaxPerPage` limit is not adjustable. You'll have to batch the requests.
If you are using the JS SDK, you can call `pb.collection("example").getFullList()`.

**Reply by dindinet** [reply]
'At the moment the MaxPerPage limit is not adjustable.' 

Does this mean that it may change in the future? Could it be a setting in the Admin Interface? 

Batching would be painful from Hugo template, but not impossible.

**Reply by ganigeorgiev** [reply]
> Does this mean that it may change in the future? Could it be a setting in the Admin Interface?

Yes, it is in my todo, but for now remains out of the scope as [other tasks](https://github.com/orgs/pocketbase/projects/2) are with higher priority and I'm already behind the initial scehdule of my other open source project.

> Batching would be painful from Hugo template, but not impossible.

Ideally you'll be able to create a helper function that will send `GET /api/colections/:collection/records` requests until `page = totalPages`, but I have no experience with Hugo and I'm not sure what is the recommended approach.

_Keep in mind that fetching all records in a single HTTP request may not always be preferable, since it could cause OOM errors in spiky scenarios (currently the json marshalization is one of the main memory hog in PocketBase)._

**Reply by dindinet** [reply]
My use case is pocketbase as a CMS used by a small team of 1-5.

Fetching of content from the api would only be necessary when a website build is triggered (10, 20 times a day). 

Is it the **size** of the returned JSON that could cause the **OOM** problems or the **marshaling** of the json?

(Love the new View Collections)

**Reply by ganigeorgiev** [reply]
> Is it the size of the returned JSON that could cause the OOM problems or the marshaling of the json?

Both. If you have a large dataset (eg. 1GB+) and you want to return everything in one request it could easily drain your available memory. There is a [`GOMEMLIMIT`](https://pkg.go.dev/runtime@master#hdr-Environment_Variables) env variable that defines a soft limit and could help but it could also hurt the overall app performance since the GC will have to run more often in attempt to keep the memory consumption under the specified threshold.

**Reply by dindinet** [reply]
Oh not a worry at the moment. All our data is less than 10MB.

**Comment by mdnahas** [comment]
I just want to say this would be useful to me.  And, if it doesn't do anything, please get rid of it.

I'm seeing performance issues in a production system and thought that upping the batch/perPage value would be a quick fix.  I was shocked it did nothing!   I spent time finding it and then debugging why it wasn't working.  

I understand that performance tradeoffs exist, but having an option that does nothing when you increase is it was an unneeded distraction and false hope in a time when I'm stressed.

**Reply by ganigeorgiev** [reply]
@mdnahas I don't understand what do you mean by "if it doesn't do anything, please get rid of it"

The max perPage is already bumped to 1000 and it is there to prevent DoS-ing your server. There always should be a limit to how much data a single request can return and 1000 is already too much in my opinion.

Instead of vaguely complaining and demanding things, it would be more helpful to elaborate more on the problem that you've experienced, what PocketBase version you are using and why do you need more than 1k+ records in a single request, aka. what are you doing with the fetched result.

**Reply by mdnahas** [reply]
Sorry, if I was rude.  I'm a little rushed right now.

I have an order system.  The admin page is downloading about 5000 orders and about 8000 inventory records, along with various other things.  The admin page is loading slowly.  

I used the "network" tab on Firefox's Inspect page to diagnose the problem.  It looks like most of the time is spent downloading the records.  There are 26 requests of 200 to 700kB.  Each requests takes 70 to 300ms.  Most of the time in each request is spent "waiting" (vs. sending and receiving).  

I assumed the best way to speed this up was to have fewer requests.  That is, transfer more records per request.  So, I increased perPage to 3000.  

When I looked at the Inspect page, it was still doing the same number of requests.  Upping the value didn't change anything.   That's what I meant by "if it doesn't do anything".  

I threw this system together in 4 months.  I'm sure there are better ways to do things, such as do searches for orders on the server, not on the client.  I'm just trying to hold this system together for 2 more weeks, until this project is over.

**Reply by ganigeorgiev** [reply]
> I assumed the best way to speed this up was to have fewer requests. That is, transfer more records per request. So, I increased perPage to 3000.

With PocketBase v0.23+ the max `perPage` value is 1000 (_for <= v0.22 the max `perPage` is 500_) and everything above that will be capped to that limit. `getFullList` intentionally sends the batched `getList` requests one by one and not concurrently in order to minimize user DoS-ing their server and because we rely on the last underlying `getList` result whether to continue fetching or not.

Without a code sample to see how and when the requests are submitted is difficult to analyze performance issues, but if both requests to fetch the 5000 orders and 8000 inventories are executed on the same page and sequentiually/`await`-ing, then you can try to wrap the 2 `getFullList` calls in `Promise.all` to send and process them concurrently if not already (_their underlying `getList` calls will still be submitted one by one_).

If filtering and pagination on the server-side is not an option than I'm not sure what to suggest as an optimization in your use case but making the max `perPage` limit configurable was ruled out during the refactoring because it can be abused and can cause all sort of performance issues that could affect not just the individual request and could potentially take down the entire application (especially when `expand` is used).

**Reply by mdnahas** [reply]
Thanks for your answer.  You are always so patient and giving.

I think the answer is a redesign in version 2.0, with the features you suggest.  The admins are not viewing nor working on all 5000 orders at once.  But that will have to happen after this season is over in June.   

I'll also have to "upgrade" to Svelte 5 and SvelteKit 2.  It is going to be a busy off-season!
